File: legacy/capacitor-web/src/lib/teaWatchlistStorage.js
const STORAGE_KEY = 'deite_tea_watchlist_v1';

/**
 * @returns {{ id: string, title: string, url: string, postUrl: string, thumbnail: string, author: string, savedAt: string }[]}
 */
export function getTeaWatchlist() {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return [];
    const arr = JSON.parse(raw);
    return Array.isArray(arr) ? arr.filter((x) => x && typeof x === 'object') : [];
  } catch {
    return [];
  }
}

/** @param {ReturnType<typeof getTeaWatchlist>} items */
export function setTeaWatchlist(items) {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(items));
  } catch {
    /* quota / private mode */
  }
}

/** @param {Record<string, unknown>} item Raw tea carousel / feed item */
export function normalizeTeaItemForWatchlist(item) {
  return {
    id: String(item?.id ?? ''),
    title: typeof item?.title === 'string' ? item.title : '',
    url: typeof item?.url === 'string' ? item.url : '',
    postUrl: typeof item?.postUrl === 'string' ? item.postUrl : '',
    thumbnail: typeof item?.thumbnail === 'string' ? item.thumbnail : '',
    author: typeof item?.author === 'string' ? item.author : '',
    savedAt: new Date().toISOString(),
    source: 'tea',
  };
}

/**
 * Toggle item in watchlist by Reddit post id. Returns the new list.
 * @param {Record<string, unknown>} item
 */
export function toggleTeaWatchlistItem(item) {
  const id = String(item?.id ?? '');
  if (!id) return getTeaWatchlist();
  const list = getTeaWatchlist();
  const idx = list.findIndex((x) => String(x.id) === id);
  if (idx >= 0) {
    list.splice(idx, 1);
  } else {
    const row = normalizeTeaItemForWatchlist(item);
    if (!row.url) return list;
    list.unshift(row);
  }
  setTeaWatchlist(list);
  return list;
}

export function removeTeaWatchlistById(postId) {
  const id = String(postId);
  const list = getTeaWatchlist().filter((x) => String(x.id) !== id);
  setTeaWatchlist(list);
  return list;
}

export function isTeaPostWatchlisted(postId) {
  const id = String(postId);
  return getTeaWatchlist().some((x) => String(x.id) === id);
}

File: flutter_app/lib/utils/tea_watchlist_storage.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

const _storageKey = 'deite_tea_watchlist_v1';

class TeaWatchlistItem {
  TeaWatchlistItem({
    required this.id,
    required this.title,
    required this.url,
    this.postUrl = '',
    this.thumbnail = '',
    this.author = '',
    this.savedAt = '',
    this.source = 'tea',
  });

  final String id;
  final String title;
  final String url;
  final String postUrl;
  final String thumbnail;
  final String author;
  final String savedAt;
  final String source;

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'url': url,
        'postUrl': postUrl,
        'thumbnail': thumbnail,
        'author': author,
        'savedAt': savedAt,
        'source': source,
      };

  factory TeaWatchlistItem.fromJson(Map<String, dynamic> json) => TeaWatchlistItem(
        id: '${json['id'] ?? ''}',
        title: json['title'] as String? ?? '',
        url: json['url'] as String? ?? '',
        postUrl: json['postUrl'] as String? ?? '',
        thumbnail: json['thumbnail'] as String? ?? '',
        author: json['author'] as String? ?? '',
        savedAt: json['savedAt'] as String? ?? '',
        source: json['source'] as String? ?? 'tea',
      );
}

Future<List<TeaWatchlistItem>> getTeaWatchlist() async {
  final prefs = await SharedPreferences.getInstance();
  final raw = prefs.getString(_storageKey);
  if (raw == null) return [];
  try {
    final arr = jsonDecode(raw) as List<dynamic>;
    return arr
        .whereType<Map<String, dynamic>>()
        .map(TeaWatchlistItem.fromJson)
        .toList();
  } catch (_) {
    return [];
  }
}

Future<void> setTeaWatchlist(List<TeaWatchlistItem> items) async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.setString(_storageKey, jsonEncode(items.map((e) => e.toJson()).toList()));
}

TeaWatchlistItem normalizeTeaItemForWatchlist(Map<String, dynamic> item) {
  return TeaWatchlistItem(
    id: '${item['id'] ?? ''}',
    title: item['title'] as String? ?? '',
    url: item['url'] as String? ?? '',
    postUrl: item['postUrl'] as String? ?? '',
    thumbnail: item['thumbnail'] as String? ?? '',
    author: item['author'] as String? ?? '',
    savedAt: DateTime.now().toUtc().toIso8601String(),
    source: 'tea',
  );
}

Future<List<TeaWatchlistItem>> toggleTeaWatchlistItem(Map<String, dynamic> item) async {
  final id = '${item['id'] ?? ''}';
  if (id.isEmpty) return getTeaWatchlist();
  final list = await getTeaWatchlist();
  final idx = list.indexWhere((x) => x.id == id);
  if (idx >= 0) {
    list.removeAt(idx);
  } else {
    final row = normalizeTeaItemForWatchlist(item);
    if (row.url.isEmpty) return list;
    list.insert(0, row);
  }
  await setTeaWatchlist(list);
  return list;
}

Future<List<TeaWatchlistItem>> removeTeaWatchlistById(String postId) async {
  final list = (await getTeaWatchlist()).where((x) => x.id != postId).toList();
  await setTeaWatchlist(list);
  return list;
}

Future<bool> isTeaPostWatchlisted(String postId) async {
  final list = await getTeaWatchlist();
  return list.any((x) => x.id == postId);
}

bool isDirectImageUrl(String? postUrl) {
  if (postUrl == null || postUrl.trim().isEmpty) return false;
  final path = postUrl.trim().split('?').first.split('#').first;
  return RegExp(r'\.(jpe?g|png|gif|webp)$', caseSensitive: false).hasMatch(path);
}

String? watchlistHeroUrl(TeaWatchlistItem row) {
  final thumb = row.thumbnail.trim();
  if (isDirectImageUrl(row.postUrl)) return row.postUrl.trim();
  if (thumb.startsWith('http')) return thumb;
  return null;
}
